AdaptiveLogSoftmaxWithLoss no_batch_dim support

george-qi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69054

Differential Revision: [D33200166](https://our.internmc.facebook.com/intern/diff/D33200166)

george-qi committed:
AdaptiveLogSoftmaxWithLoss no_batch_dim support

[ghstack-poisoned]

george-qi commented on torch/csrc/api/src/nn/modules/adaptive.cpp:
```diff
@@ -61,14 +61,40 @@ void AdaptiveLogSoftmaxWithLossImpl::reset_parameters() {
+  } else {
+    Tensor new_input = input;
+    Tensor new_target = target;
+  }
```
Basically just created a new array if the input/target are 1D and 0D, respectively

Mainly did this because target_mask.nonzero() will return a (1, 0) tensor if target is a scalar, but this creates extra space.. not sure if this performance hit is acceptable

jbschlosser replied:
I think the unsqueeze / re-squeeze approach is fine here. Note that unsqueeze just creates a view of the same underlying storage, so this shouldn't be too much of a perf hit

george-qi commented on torch/csrc/api/src/nn/modules/adaptive.cpp:
```diff
@@ -106,14 +132,18 @@ ASMoutput AdaptiveLogSoftmaxWithLossImpl::forward(const Tensor& input, const Ten
+  if (targ_dim == 0) {
+    output = output.squeeze(0);
+  }
+
```
Do we want the output to be a scalar here? This produces an output that looks something like:

`tensor(-4.5369, grad_fn=<SqueezeBackward1>)`

whereas the same input that isn't 1D/0D yields 

`tensor([-4.5369], grad_fn=<AddBackward0>)`

Is this okay?

jbschlosser replied:
Yeah, we want a tensor scalar output when a tensor scalar is passed in for the target.

jbschlosser commented on test/test_nn.py:
```diff
@@ -11351,6 +11351,14 @@ def test_adaptive_log_softmax(self):
 
         self.assertEqual(asfm(x, y).output.size(), (2, ))
 
+        # test no_batch_dim support
+        asfm = nn.AdaptiveLogSoftmaxWithLoss(16, 20, [5, 10, 15], div_value=2.)
+        x = torch.randn(1, 16)
+        y = torch.tensor([17])
+        x2 = x.squeeze(0)
+        y2 = y.squeeze(0)
+        self.assertEqual(asfm(x, y).output.squeeze(0), asfm(x2, y2).output)
```
Thanks for adding the tests! Note that there is some precedent for using a reference function alongside the old style dictionary-based tests (see [here](https://github.com/pytorch/pytorch/blob/251686fc4cb1962944ed99c938df2d54f3d62e46/torch/testing/_internal/common_nn.py#L1315-L1329) for an example).

It won't work directly here because this module returns multiple outputs, but perhaps this can be made a bit more consistent with what's been done before

jbschlosser commented on torch/nn/modules/adaptive.py:
```diff
@@ -93,9 +93,9 @@ class AdaptiveLogSoftmaxWithLoss(Module):
     Shape:
-        - input: :math:`(N, \texttt{in\_features})`
+        - input: :math:`(N, \texttt{in\_features})` or :math:`(\texttt{in\_features})`
         - target: :math:`(N)` where each value satisfies :math:`0 <= \texttt{target[i]} <= \texttt{n\_classes}`
```
Note that target should be of shape ``` :math:`N` ``` or ``` :math:`()` ``` depending on the input

jbschlosser commented on torch/csrc/api/src/nn/modules/adaptive.cpp:
```diff
@@ -61,14 +61,40 @@ void AdaptiveLogSoftmaxWithLossImpl::reset_parameters() {
+      ".");
+  }
+
+  Tensor new_input, new_target;
```
nit: the usual style is to use names like `input_` / `target_` for the args. Then you can do:

```cpp
bool is_batched = (target_dim > 0);
Tensor input = is_batched ? input_.unsqueeze(0) : input_;
Tensor target = is_batched ? target_.unsqueeze(0) : target_;
```

jbschlosser commented on torch/csrc/api/src/nn/modules/adaptive.cpp:
```diff
@@ -106,14 +132,18 @@ ASMoutput AdaptiveLogSoftmaxWithLossImpl::forward(const Tensor& input, const Ten
   output += head_logprob.gather(1, gather_inds.unsqueeze(1)).squeeze();
   const double loss = (-output).mean().item().toDouble();
 
+  if (targ_dim == 0) {
```
nit: use bool defined above named something like `is_batched`

jbschlosser commented on torch/nn/modules/adaptive.py:
```diff
@@ -167,34 +167,54 @@ def reset_parameters(self) -> None:
-        if input.size(0) != target.size(0):
-            raise RuntimeError('Input and target should have the same size '
-                               'in the batch dimension.')
+        targ_dim = target.dim()
+
+        if targ_dim == 1:
+            if input.size(0) != target.size(0):
+                raise RuntimeError('Input and target should have the same size '
+                                   'in the batch dimension.')
+            if input.dim() != 2:
+                raise RuntimeError('1D target tensor expects 2D input tensors, ' 
+                                   'but found inputs with size', input.size())
+        elif targ_dim == 0:
+            if input.dim() != 1:
+                raise RuntimeError('0D target tensor expects 1D input tensors, ' 
+                                   'but found inputs with size', input.size())
+        else:
+            raise RuntimeError('0D or 1D target tensor expected, ' 
+                               'multi-target not supported')
+
+        if targ_dim == 0:
+            new_input = input.unsqueeze(0)
+            new_target = target.unsqueeze(0)
+        else:
+            new_input = input
+            new_target = target
```
Dang, they duplicated this logic across python and C++? I thought only MultiHeadAttention was like that :p

Similar comments as above then apply here as well

jbschlosser commented on torch/csrc/api/src/nn/modules/adaptive.cpp:
```diff
@@ -61,14 +61,40 @@ void AdaptiveLogSoftmaxWithLossImpl::reset_parameters() {
+  auto targ_dim = target.dim();
+
+  TORCH_CHECK(
+    targ_dim == 1 || targ_dim == 0,
```
I recommend using a well-named bool throughout (maybe `is_batched`)

jbschlosser reviewed (commented):
Thanks for tackling this! I left a few nitpicky comments below but it's looking pretty good :)

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"




[ghstack-poisoned]

jbschlosser commented on test/test_nn.py:
```diff
@@ -11348,9 +11348,17 @@ def test_adaptive_log_softmax(self):
         self.assertEqual(asfm.tail[1][1].weight.size(), (5, 4))
         self.assertEqual(asfm.tail[2][1].weight.size(), (5, 2))
-
+        print("boutta")
```
nit: leftover print :)

jbschlosser commented on torch/nn/modules/adaptive.py:
```diff
@@ -200,7 +217,7 @@ def forward(self, input: Tensor, target: Tensor) -> _ASMoutput:
 
                 gather_inds.index_fill_(0, row_indices, cluster_index)
-
+                
```
nit: need to delete the whitespace to satisfy the linter

jbschlosser reviewed (commented):
I think we'll also need some tests to verify the logic on the C++ side (both functional and module forms) since it's duplicated

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"




[ghstack-poisoned]

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"




[ghstack-poisoned]

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"




[ghstack-poisoned]

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"




[ghstack-poisoned]

jbschlosser reviewed (commented):
Looking pretty good, just need a test or two on the C++ side for the C++ module changes :)

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33200166).

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"


Differential Revision: [D33200166](https://our.internmc.facebook.com/intern/diff/D33200166)

[ghstack-poisoned]

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33200166).

jbschlosser reviewed (approved):
LGTM! Thanks :)

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"


Differential Revision: [D33200166](https://our.internmc.facebook.com/intern/diff/D33200166)

[ghstack-poisoned]

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"


Differential Revision: [D33200166](https://our.internmc.facebook.com/intern/diff/D33200166)

[ghstack-poisoned]

george-qi committed:
Update on "AdaptiveLogSoftmaxWithLoss no_batch_dim support"


Differential Revision: [D33200166](https://our.internmc.facebook.com/intern/diff/D33200166)

[ghstack-poisoned]

george-qi commented:
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33200166).
